// @azure/ai-text-analytics/dist-esm/src/generated/models/parameters.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { MultiLanguageBatchInput as MultiLanguageBatchInputMapper, LanguageBatchInput as LanguageBatchInputMapper } from "../models/mappers";
export const contentType = {
    parameterPath: ["options", "contentType"],
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Content-Type",
        type: {
            name: "String"
        }
    }
};
export const input = {
    parameterPath: "input",
    mapper: MultiLanguageBatchInputMapper
};
export const endpoint = {
    parameterPath: "endpoint",
    mapper: {
        serializedName: "Endpoint",
        required: true,
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export const modelVersion = {
    parameterPath: ["options", "modelVersion"],
    mapper: {
        serializedName: "model-version",
        type: {
            name: "String"
        }
    }
};
export const includeStatistics = {
    parameterPath: ["options", "includeStatistics"],
    mapper: {
        serializedName: "includeStatistics",
        type: {
            name: "Boolean"
        }
    }
};
export const stringIndexType = {
    parameterPath: ["options", "stringIndexType"],
    mapper: {
        defaultValue: "TextElements_v8",
        serializedName: "stringIndexType",
        type: {
            name: "String"
        }
    }
};
export const domain = {
    parameterPath: ["options", "domain"],
    mapper: {
        serializedName: "domain",
        type: {
            name: "String"
        }
    }
};
export const input1 = {
    parameterPath: "input",
    mapper: LanguageBatchInputMapper
};
export const opinionMining = {
    parameterPath: ["options", "opinionMining"],
    mapper: {
        serializedName: "opinionMining",
        type: {
            name: "Boolean"
        }
    }
};
